### PR TITLE
feat: add regional support

### PR DESCRIPTION
**Pull request by thinkingserious**
Adds Twilio regional support.

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by thinkingserious** [comment]
Not sure why the tests are failing, locally I get:

```bash
$ npm test

> twilio@3.42.2 test ./twilio-node/di-640/twilio-node
> npm run test:javascript && npm run test:typescript


> twilio@3.42.2 test:javascript ./twilio-node/di-640/twilio-node
> jasmine spec/**/**.spec.js

Randomized with seed 21297
Started
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................


3345 specs, 0 failures
Finished in 9.425 seconds
Randomized with seed 21297 (jasmine --random=true --seed=21297)

> twilio@3.42.2 test:typescript ./twilio-node/di-640/twilio-node
> tsc examples/typescript/example.ts --noEmit --strict
```
Will investigate further.

File: lib/rest/Twilio.d.ts [hunk 1]
```diff
@@ -146,13 +146,15 @@ declare namespace Twilio {
    * Options to pass to the Twilio Client constructor
    *
    * @property accountSid - The default accountSid. This is set to username if not provided
+   * @property edge - Twilio region to use. Defaults to none
```
Comment (childish-sambino) [inline lib/rest/Twilio.d.ts:149]: `s/region/edge/`

File: lib/rest/Twilio.js [hunk 2]
```diff
@@ -102,8 +103,9 @@ var RestException = require('../base/RestException');  /* jshint ignore:line */
  *          The client used for http requests. Defaults to RequestClient
  * @param {string} [opts.accountSid] -
  *          The default accountSid. This is set to username if not provided
+ * @param {string} [opts.edge] - Twilio region to use. Defaults to none
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:106]: Group this with the region docs.

File: lib/rest/Twilio.js [hunk 3]
```diff
@@ -10,6 +10,7 @@
 /* jshint ignore:end */
 
 var moduleInfo = require('../../package.json');  /* jshint ignore:line */
+var url = require('url');  /* jshint ignore:line */
 var _ = require('lodash');  /* jshint ignore:line */
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:14]: Can we drop lodash from this now?

File: lib/rest/Twilio.js [hunk 4]
```diff
@@ -102,8 +103,9 @@ var RestException = require('../base/RestException');  /* jshint ignore:line */
  *          The client used for http requests. Defaults to RequestClient
  * @param {string} [opts.accountSid] -
  *          The default accountSid. This is set to username if not provided
+ * @param {string} [opts.edge] - Twilio region to use. Defaults to none
  * @param {object} [opts.env] - The environment object. Defaults to process.env
- * @param {string} [opts.region] - Twilio region to use. Defaults to none
+ * @param {string} [opts.region] - Twilio region to use. Defaults to defaultRegion
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:108]: Only defaults to `us1` if an edge is provided.

File: lib/rest/Twilio.js [hunk 5]
```diff
@@ -249,18 +253,17 @@ Twilio.prototype.request = function request(opts) {
...
-  if (this.region) {
-    var parts = _.split(uri, '.');
-
-    if (parts.length > 1 && !_.isEqual(parts[1], this.region)) {
-      uri = _.join(_.concat([parts[0], this.region], _.slice(parts, 1)), '.');
-    }
+  var uri = new url.URL(opts.uri);
+  var [product, subdomain, domain] = uri.hostname.split('.');
+  if (this.edge) {
+    uri.hostname = ''.concat(product, '.', this.edge, '.', this.region || defaultRegion, '.', subdomain, '.', domain);
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:259]: Nice.

File: lib/rest/Twilio.js [hunk 6]
```diff
@@ -249,18 +253,17 @@ Twilio.prototype.request = function request(opts) {
...
+  } else if (this.region) {
+    uri.hostname = ''.concat(product, '.', this.region, '.', subdomain, '.', domain);
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:261]: Should just concat on `'.'`

File: lib/rest/Twilio.js [hunk 7]
```diff
@@ -249,18 +253,17 @@ Twilio.prototype.request = function request(opts) {
...
 
-  var uri = opts.uri;
-  if (this.region) {
-    var parts = _.split(uri, '.');
-
-    if (parts.length > 1 && !_.isEqual(parts[1], this.region)) {
-      uri = _.join(_.concat([parts[0], this.region], _.slice(parts, 1)), '.');
-    }
+  var uri = new url.URL(opts.uri);
+  var [product, subdomain, domain] = uri.hostname.split('.');
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:257]: I _think_ it's possible the uri is already going to contain an edge/region. Think pagination/links.

File: spec/unit/rest/Twilio.spec.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+var proxyquire = require('proxyquire');
+var url = require('url');  /* jshint ignore:line */
+var Q = require('q');
+
+describe('hostname', function() {
+  var client;
+  beforeEach(function() {
```
Comment (childish-sambino) [inline spec/unit/rest/Twilio.spec.js:7]: Can we use `nock` instead of a mock here? Recent example: https://github.com/sendgrid/sendgrid-nodejs/pull/1093/files#diff-0a22ccfa1c28c602273fbc87ebe73ac2R42

**Comment by childish-sambino** [comment]
Also need examples/docs.

File: lib/rest/Twilio.js (hunk 3 above)
Comment (thinkingserious) [reply lib/rest/Twilio.js:14]: Yes

File: lib/rest/Twilio.js (hunk 6 above)
Comment (thinkingserious) [reply lib/rest/Twilio.js:261]: Do you suggest using a `join` for this?
e.g.
```javascript
var hostname = [product, edge, region, subdomain, domain]
hostname.join('.')
```

Comment (childish-sambino) [reply lib/rest/Twilio.js:261]: Yes.

File: lib/rest/Twilio.js [hunk 9]
```diff
@@ -249,18 +252,44 @@ Twilio.prototype.request = function request(opts) {
...
+  if (parsedHostname.length === 5) {
+    [product, edge, region, subdomain, domain] = uri.hostname.split('.');
+    if (this.edge) {
+      uri.hostname = [product, this.edge || edge, this.region || region, subdomain, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.edge || edge, this.region, subdomain, domain].join('.');
+    }
+  }
+  // https://product.region.subdomain.domain
+  if (parsedHostname.length === 4) {
```
Comment (eshanholtz) [inline lib/rest/Twilio.js:272]: `else if (parsedHostname.length === 4)`

File: lib/rest/Twilio.js [hunk 10]
```diff
@@ -249,18 +252,44 @@ Twilio.prototype.request = function request(opts) {
...
+    [product, region, subdomain, domain] = uri.hostname.split('.');
+    if (this.edge) {
+      uri.hostname = [product, this.edge, this.region || region, subdomain, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.region || region, subdomain, domain].join('.');
+    }
+  }
+  // https://product.subdomain.domain
+  if (parsedHostname.length === 3) {
```
Comment (eshanholtz) [inline lib/rest/Twilio.js:281]: else if (parsedHostname.length === 3)

File: lib/rest/Twilio.d.ts (hunk 1 above)
Comment (childish-sambino) [reply lib/rest/Twilio.d.ts:149]: This has not been resolved.

File: lib/rest/Twilio.js (hunk 2 above)
Comment (childish-sambino) [reply lib/rest/Twilio.js:105]: This has not been resolved.

File: lib/rest/Twilio.js (hunk 4 above)
Comment (childish-sambino) [reply lib/rest/Twilio.js:107]: This has not been resolved.

File: lib/rest/Twilio.js [hunk 11]
```diff
@@ -249,18 +252,39 @@ Twilio.prototype.request = function request(opts) {
...
-    if (parts.length > 1 && !_.isEqual(parts[1], this.region)) {
-      uri = _.join(_.concat([parts[0], this.region], _.slice(parts, 1)), '.');
+  var product;
+  var edge;
+  var region;
+  var subdomain;
+  var domain;
+  var uri = new url.URL(opts.uri);
+  var parsedHostname = uri.hostname.split('.');
+  if (parsedHostname.length === 5) { // https://product.edge.region.subdomain.domain
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:262]: Think it's simpler to first get everything before `.twilio.com` (if it exists), then split the prefix into 3 parts (`[product, edgeOrRegion, region] = prefix.split('.')`), and then reconstruct it based on what's defined.

File: package.json [hunk 12]
```diff
@@ -41,6 +41,7 @@
     "jsdoc": "^3.6.3",
     "jshint": "^2.11.0",
     "mock-fs": "^4.11.0",
+    "nock": "^12.0.3",
```
Comment (childish-sambino) [inline package.json:44]: Version doesn't support Node v6/8: https://github.com/sendgrid/sendgrid-nodejs/pull/1093/commits/58db16c6c680fe78fca073944c6c56cc38ead002

File: spec/unit/rest/Twilio.spec.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+  describe('setting the region', () => {
+    it('should use the default region if only edge is defined', () => {
+      client.edge = 'edge';
+      client.messages.create({
+        body: 'Hello from Node',
+        to: '+XXXXXXXXXX',
+        from: '+XXXXXXXXXX',
+      });
+      var uri = new url.URL(client._httpClient.lastRequest.url);
```
Comment (childish-sambino) [inline spec/unit/rest/Twilio.spec.js:24]: Don't use `lastRequest` since the details of the underlying HTTP client could change over time. Better to use nock to mock the specific request you're expecting. See my example here: https://github.com/sendgrid/sendgrid-nodejs/pull/1093/files#diff-0a22ccfa1c28c602273fbc87ebe73ac2R41-R52

File: spec/unit/rest/Twilio.spec.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    it('should set the region and edge properly', () => {
+      client.edge = 'edge';
+      client.region = 'region';
+      client.messages.create({
+        body: 'Hello from Node',
+        to: '+XXXXXXXXXX',
+        from: '+XXXXXXXXXX',
+      });
+      var uri = new url.URL(client._httpClient.lastRequest.url);
+      expect(uri.hostname).toEqual('api.edge.region.twilio.com');
```
Comment (childish-sambino) [inline spec/unit/rest/Twilio.spec.js:56]: Need tests that pass a URI with edge/region included.

File: lib/rest/Twilio.d.ts (hunk 1 above)
Comment (thinkingserious) [reply lib/rest/Twilio.d.ts:149]: Sorry, I made these changes in Yoyodyne and have not applied them yet.

File: lib/rest/Twilio.js (hunk 11 above)
Comment (thinkingserious) [reply lib/rest/Twilio.js:262]: I'll try that.

File: lib/rest/Twilio.js [hunk 15]
```diff
@@ -249,18 +253,36 @@ Twilio.prototype.request = function request(opts) {
...
+    if (this.edge) {
+      uri.hostname = [product, this.edge || edgeOrRegion, this.region || region, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.region || edgeOrRegion, domain].join('.');
+    }
+  } else if (!edgeOrRegion && !region) { // https://product.subdomain.domain
+    if (this.edge) {
+      uri.hostname = [product, this.edge, this.region || defaultRegion, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.region, domain].join('.');
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:279]: Still overly complex. What about something like this:
```
function getHostname(uri, targetEdge, targetRegion) {
  const domain = 'twilio.com';
  const defaultRegion = 'us1';

  targetRegion = targetRegion || (targetEdge && defaultRegion);

  if (!targetRegion) {
    return uri;
  }

  const prefix = uri.split('.' + domain)[0];
  let [product, edge, region] = prefix.split('.');

  if (!region || targetEdge) {
    edge = targetEdge;
  }

  region = targetRegion;

  return [product, edge, region, domain].filter(part => part).join('.');
}
```

Comment (thinkingserious) [reply lib/rest/Twilio.js]: Updated, thanks!

File: spec/unit/rest/Twilio.spec.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+const nock = require('nock');
+var url = require('url');  /* jshint ignore:line */
+
+describe('client', () => {
+  var client;
+  const twilio = require('../../../lib');
+
+  describe('setting the region', () => {
+    it('should use the default region if only edge is defined', () => {
+      client = new twilio('ACXXXXXXXX', 'test-password');
```
Comment (childish-sambino) [inline spec/unit/rest/Twilio.spec.js:11]: This can be constructed outside the testing block. Or thrown into a `beforeEach`.

File: lib/rest/Twilio.js [hunk 17]
```diff
@@ -271,6 +268,44 @@ Twilio.prototype.request = function request(opts) {
...
+  const domain = 'twilio.com';
+  const defaultRegion = 'us1';
+  const prefix = hostname.split('.' + domain)[0];
+  let [product, edge, region] = prefix.split('.');
+
+  if (region) {
+    targetRegion = targetRegion || region;
+  }
+
+  targetRegion = targetRegion || (targetEdge && defaultRegion);
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:294]: In the case of `gethostname('api.region.twilio.com', 'edge')` it will return `api.edge.region.twilio.com` instead of `api.edge.us1.twilio.com`. Is this desired?

File: spec/unit/rest/Twilio.spec.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+      client.edge = 'edge';
+      return client.request({method: 'GET', uri: 'https://api.twilio.com'})
+        .then(() => scope.done());
+    });
+    it('should use the default region if only edge is defined and there is a defined region', () => {
+      const scope = nock('https://api.edge.us1.twilio.com')
+        .get('/')
+        .reply(200, 'test response');
+      client.edge = 'edge';
+      return client.request({method: 'GET', uri: 'https://api.region.twilio.com'})
```
Comment (childish-sambino) [inline spec/unit/rest/Twilio.spec.js:26]: Test needed where only `client.region` is set and hostname already contains a region.

File: lib/rest/Twilio.js [hunk 19]
```diff
@@ -271,6 +268,40 @@ Twilio.prototype.request = function request(opts) {
...
+ * @param {string} targetEdge - The targeted edge location (e.g. sydney)
+ * @param {string} targetRegion - The targeted region location (e.g. au1)
+ */
+/* jshint ignore:end */
+Twilio.prototype.getHostname = function getHostname(hostname, targetEdge,
+                                                     targetRegion) {
+  const domain = 'twilio.com';
+  const defaultRegion = 'us1';
+
+  const prefix = hostname.split('.' + domain)[0];
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:288]: What happens when the hostname doesn't contain `.twilio.com`?

Comment (thinkingserious) [reply lib/rest/Twilio.js:288]: Added a test to account for this case. We now have the same behavior as if the domain was `twilio.com`.

File: spec/unit/rest/Twilio.spec.js (hunk 18 above)
Comment (thinkingserious) [reply spec/unit/rest/Twilio.spec.js:26]: https://github.com/twilio/twilio-node/pull/563/files#diff-78816a2ef39228b6832a61b7014e7504R62

File: lib/rest/Twilio.js [hunk 20]
```diff
@@ -271,6 +268,40 @@ Twilio.prototype.request = function request(opts) {
...
+Twilio.prototype.getHostname = function getHostname(hostname, targetEdge,
+                                                     targetRegion) {
+  const domain = hostname.split('.').slice(-2).join('.');
+  const defaultRegion = 'us1';
+
+  const prefix = hostname.split('.' + domain)[0];
+  let [product, edge, region] = prefix.split('.');
+
+  targetRegion = targetRegion || (targetEdge && defaultRegion);
+  if (!targetRegion) {
```
Comment (childish-sambino) [inline lib/rest/Twilio.js:292]: Should move this check to before the hostname parsing logic. No need to parse the hostname if we're not modifying it.
